iml-action-client/src/lib.rs

```diff
@@ -2,35 +2,75 @@
 // Use of this source code is governed by a MIT-style
 // license that can be found in the LICENSE file.
 
+use bytes::buf::BufExt as _;
 use futures::{
     channel::oneshot,
     future::{self, Either},
     Future, FutureExt, TryFutureExt,
 };
+use hyper::{client::HttpConnector, Body, Client, Request};
+use hyperlocal::{UnixClientExt as _, UnixConnector};
 use iml_action_runner::ActionType;
-use iml_manager_env::get_action_runner_connect;
+use iml_manager_env::{get_action_runner_http, get_action_runner_uds, running_in_docker};
 use iml_wire_types::{Action, ActionId, ActionName, Fqdn};
 use thiserror::Error;
 use uuid::Uuid;
 
 #[derive(Error, Debug)]
 pub enum ImlActionClientError {
     #[error(transparent)]
-    ReqwestError(#[from] reqwest::Error),
+    HyperError(#[from] hyper::Error),
+    #[error(transparent)]
+    UriError(hyper::http::uri::InvalidUri),
     #[error("Request Cancelled")]
     CancelledRequest,
+    #[error(transparent)]
+    HttpError(#[from] hyper::http::Error),
+    #[error(transparent)]
+    SerdeJsonError(#[from] serde_json::error::Error),
 }
 
-pub fn invoke_rust_agent(
-    host: impl Into<Fqdn> + Clone,
-    command: impl Into<ActionName>,
-    args: impl serde::Serialize,
-) -> (
-    oneshot::Sender<()>,
-    impl Future<Output = Result<serde_json::Value, ImlActionClientError>>,
-) {
-    let request_id = Uuid::new_v4().to_hyphenated().to_string();
-    let conn = get_action_runner_connect();
+#[derive(Clone)]
+enum ClientWrapper {
+    Http(Client<HttpConnector>),
+    Unix(Client<UnixConnector>),
+}
+
+impl ClientWrapper {
+    fn request(&self, req: Request<Body>) -> hyper::client::ResponseFuture {
+        match self {
+            ClientWrapper::Http(c) => c.request(req),
+            ClientWrapper::Unix(c) => c.request(req),
+        }
+    }
+}
+
+fn client() -> ClientWrapper {
+    if running_in_docker() {
+        ClientWrapper::Http(Client::new())
+    } else {
+        ClientWrapper::Unix(Client::unix())
+    }
+}
+
+fn connect_uri() -> Result<hyper::Uri, ImlActionClientError> {
+    if running_in_docker() {
+        get_action_runner_http()
```

Comment:
This is going to do a env lookup each for each connection, and then parse it to a Uri. this fn should instead return a Uri directly or panic if it can't be converted.